mono/utils/mono-log-common.c

```diff
@@ -0,0 +1,157 @@
+/*
+ * mono-log-common.c: Platform-independent interface to the logger
+ *
+ * This module contains the POSIX syslog logger interface
+ *
+ * Author:
+ *    Neale Ferguson <neale@sinenomine.net>
+ *
+ */
+#include <config.h>
+
+#ifdef HAVE_UNISTD_H
+#include <unistd.h>
+#endif
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <ctype.h>
+#include <string.h>
+#include <glib.h>
+#include <errno.h>
+#include <time.h>
+#ifndef HOST_WIN32
+#include <sys/time.h>
+#else
+#include <process.h>
+#endif
+#include "mono-logger.h"
+
+static FILE *logFile = NULL;
+static void *logUserData = NULL;
+
+/**
+ * mapSyslogLevel:
+ * 	
+ * 	@level - GLogLevelFlags value
+ * 	@returns The equivalent character identifier
+ */
+static inline char 
+mapLogFileLevel(GLogLevelFlags level) 
+{
+	if (level & G_LOG_LEVEL_ERROR)
+		return ('E');
+	if (level & G_LOG_LEVEL_CRITICAL)
+		return ('C');
+	if (level & G_LOG_LEVEL_WARNING)
+		return ('W');
+	if (level & G_LOG_LEVEL_MESSAGE)
+		return ('N');
+	if (level & G_LOG_LEVEL_INFO)
+		return ('I');
+	if (level & G_LOG_LEVEL_DEBUG)
+		return ('D');
+	return ('I');
+}
+
+/**
+ * mono_log_open_logfile
+ * 	
+ *	Open the logfile. If the path is not specified default to stdout. If the
+ *	open fails issue a warning and use stdout as the log file destination.
+ *
+ * 	@path - Path for log file
+ * 	@userData - Not used
+ */
+void
+mono_log_open_logfile(const char *path, void *userData)
+{
+	if (path == NULL) {
+		logFile = stdout;
+	} else {
+#ifndef HOST_WIN32
+		logFile = fopen(path, "w");
+#else
+		gunichar2 *wPath = g_utf8_to_utf16(path, -1, 0, 0, 0);
+		if (wPath != NULL) {
+			logFile = _wfopen((wchar_t *) wPath, L"w");
+			g_free (wPath);
+		}
+#endif
+		if (logFile == NULL) {
+			g_warning("opening of log file %s failed with %s - defaulting to stdout", 
+				  path, strerror(errno));
+			logFile = stdout;
+		}
+	}
+	logUserData = userData;
+}
+
+/**
+ * mono_log_write_logfile
+ * 	
+ * 	Write data to the log file.
+ *
+ * 	@domain - Identifier string
+ * 	@level - Logging level flags
+ * 	@format - Printf format string
+ * 	@vargs - Variable argument list
+ */
+void
+mono_log_write_logfile(const char *domain, GLogLevelFlags level, mono_bool hdr, const char *format, va_list args)
+{
+	time_t t;
+	char logTime[80],	
+	     logMessage[512];
+	pid_t pid;
+	int iLog = 0;
+	size_t nLog;
+
+	if (logFile == NULL)
+		logFile = stdout;
+
+	if (hdr) {
+#ifndef HOST_WIN32
+		struct tm tod;
+		time(&t);
+		localtime_r(&t, &tod);
+		pid = getpid();
+		strftime(logTime, sizeof(logTime), "%F %T", &tod);
+#else
+		struct tm *tod;
+		time(&t);
+		tod = localtime(&t);
+		pid = _getpid();
+		strftime(logTime, sizeof(logTime), "%F %T", tod);
+#endif
+		iLog = snprintf(logMessage, sizeof(logMessage), "%s level[%c] mono[%d]: ",
+				logTime,mapLogFileLevel(level),pid);
+	}
+	nLog = sizeof(logMessage) - iLog - 2;
+	iLog = vsnprintf(logMessage+iLog, nLog, format, args);
+#ifdef HOST_WIN32
+	logMessage[iLog++] = '\r';
```

Comment:
Will remove from common and windows.

From: Johan Lorensson +ADw-notifications+AEA-github.com+ADw-mailto:notifications+AEA-github.com+AD4APg-
Reply-To: mono/mono +ADw-reply++0004b6fadc5ddbab691d4375d17b0c0d57e43496f536ac8692cf00000001136962e692a170ce03eae4AQA-reply.github.com+ADw-mailto:reply++0004b6fadc5ddbab691d4375d17b0c0d57e43496f536ac8692cf00000001136962e692a170ce03eae4AQA-reply.github.com+AD4APg-
Date: Friday, June 3, 2016 at 11:23 AM
To: mono/mono +ADw-mono+AEA-noreply.github.com+ADw-mailto:mono+AEA-noreply.github.com+AD4APg-
Cc: Neale Ferguson +ADw-neale+AEA-sinenomine.net+ADw-mailto:neale+AEA-sinenomine.net+AD4APg-, Mention +ADw-mention+AEA-noreply.github.com+ADw-mailto:mention+AEA-noreply.github.com+AD4APg-
Subject: Re: +AFs-mono/mono+AF0- Enhance log tracing (+ACM-2734)


In mono/utils/mono-log-common.c+ADw-https://github.com/mono/mono/pull/2734+ACM-discussion+AF8-r65725717+AD4-:

+AD4-               pid +AD0- getpid()+ADs-
+AD4-               strftime(logTime, sizeof(logTime), +ACIAJQ-F +ACU-T+ACI-, +ACY-tod)+ADs-
+AD4- +ACM-else
+AD4-               struct tm +ACo-tod+ADs-
+AD4-               time(+ACY-t)+ADs-
+AD4-               tod +AD0- localtime(+ACY-t)+ADs-
+AD4-               pid +AD0- +AF8-getpid()+ADs-
+AD4-               strftime(logTime, sizeof(logTime), +ACIAJQ-F +ACU-T+ACI-, tod)+ADs-
+AD4- +ACM-endif
+AD4-               iLog +AD0- snprintf(logMessage, sizeof(logMessage), +ACIAJQ-s level+AFsAJQ-c+AF0- mono+AFsAJQ-d+AF0-: +ACI-,
+AD4-                               logTime,mapLogFileLevel(level),pid)+ADs-
+AD4-       +AH0-
+AD4-       nLog +AD0- sizeof(logMessage) - iLog - 2+ADs-
+AD4-       iLog +AD0- vsnprintf(logMessage++iI-, nLog, format, args)+ADs-
+AD4- +ACM-ifdef HOST+AF8-WIN32
+AD4-       logMessage+AFs-iLog+++8AXQ- +AD0- '+AFw-r'+ADs-


I removed this comment as I wrote it since I realized you used the file stream in text mode and then c-runtime replace +AFw-n to +AFw-r+AFw-n behind the scenes when just using +AFw-n.

+IBQ-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub+ADw-https://github.com/mono/mono/pull/2734/files/c41ea7babeb0811068f565a7d0262b6305bf0e32+ACM-r65725717+AD4-, or mute the thread+ADw-https://github.com/notifications/unsubscribe/AAS2-oYIURE0YF5yn1SaPeqEG2Poqf8xks5qIEbmgaJpZM4HsGtY+AD4-.